### PR TITLE
New-NsxSecurityPolicy

### PR DESCRIPTION
**Pull request by Virtualizestuff**
Contains the additional internal functions New-NsxSecurityPolicyFirewallRuleSpec and New-NsxSecurityPolicyGISSpec.

### CODE REVIEW

**Comment by nmbradford** [comment]
Hi Dave - thanks so much for these PRs.  There is a lot to look through, but Im determined not to let them rot this time!  

You may have noticed we have just implemented CI integration that allows us to test PRs, to that end, it would be great if you are able to write some tests to go with these? (They allow you to be more certain that future changes to your functions dont break things in unexpected or untracked ways, so you stand to benefit :) ) 

Let me know if you would like some help with this - you can start by checking out https://powernsx.github.io/contrib/

File: PowerNSX.psm1 [hunk 1]
```diff
@@ -27578,6 +27578,383 @@ function Get-NsxSecurityPolicy {
...
+function New-NsxSecurityPolicyFirewallRuleSpec {
+    <#
+    .SYNOPSIS
+    An internal function that creates Security Policy Firewall Rules in the appropriate XML format.
+
+    .DESCRIPTION
+    This cmdlet has serveral parameters that are required to successfully create a firewall rule in a Security Policy. The information passed into the function are then processed to generate a proper XML variable which is then passed to the New-NsxSecurityPolicy function.
+
+    .EXAMPLE
+    C:\> New-NsxSecurityPolicyFirewallRuleSpec -Name "Block Web to Demo VM" -Description "Block Inbound Web traffic" -Service (Get-NsxService HTTP), (Get-NsxService HTTPS) -securityGroup (Get-NsxSecurityGroup | ?{$_.name -eq "SG App Servers"}) -Direction inbound -Enabled $true -Logging $true -Action block | Format-XML
```
Comment (alagoutte) [inline PowerNSX.psm1]: no need C:\> before the command

File: PowerNSX.psm1 [hunk 2]
```diff
@@ -27578,6 +27578,383 @@ function Get-NsxSecurityPolicy {
...
+    <#
+    .SYNOPSIS
+    An internal function that creates Security Policy Guest Introspection Service Rules in the appropriate XML format.
+    
+    .DESCRIPTION
+    This cmdlet has serveral parameters that are required to successfully create a guest introspection service rule in a Security Policy. 
+    The information passed into the function are then processed to generate a proper XML variable which
+    is then passed on to the New-NsxSecurityPolicy function 
+
+    .EXAMPLE
```
Comment (alagoutte) [inline PowerNSX.psm1]: Add a example ?

File: PowerNSX.psm1 [hunk 3]
```diff
@@ -27578,6 +27578,383 @@ function Get-NsxSecurityPolicy {
...
+                                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action/secondarySecurityGroup[$count]") -xmlElementName "objectId" -xmlElementText $member.objectId
+                            } else {
+                                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action/secondarySecurityGroup[$count]") -xmlElementName "objectId" -xmlElementText $member.objectId
+                      
+                            }
+                            $count++
+                        }
+                }
+    
+                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action")  -xmlElementName "applications"
```
Comment (dcoghlan) [inline PowerNSX.psm1]: This line needs to be executed only if a service is defined. If no service is defined (i.e. Any) then it still creates the element but with no applications, which causes the rule to be marked as "invalidApplications = true".

Comment (Virtualizestuff) [inline PowerNSX.psm1]: Thanks @dcoghlan, I'll look into this today. Appreciate the feedback!

File: PowerNSX.psm1 (hunk 2 above)
Comment (Virtualizestuff) [inline PowerNSX.psm1]: Added an example in latest commit. Thanks @alagoutte

File: PowerNSX.psm1 (hunk 3 above)
Comment (Virtualizestuff) [inline PowerNSX.psm1]: @dcoghlan I fixed the invalidApplications bug thanks for catching that!

**Pull request merged**